### PR TITLE
logind: allow IdleAction to send Lock event when it is set

### PR DESCRIPTION
**Pull request by venkata-pyla**
as per the man page[1] for IdleAction=Lock the systemd should
send the lock event which is may be registered by some application
to recieve the event and lock the login session.

same was reported in[2]

[1]https://www.freedesktop.org/software/systemd/man/logind.conf.html#IdleAction=
[2]https://github.com/systemd/systemd/issues/16391

### CODE REVIEW

**Comment by venkata-pyla** [comment]
Hi,

can i know the feedback for this issue.

thanks
venkat

**Comment by poettering** [comment]
Sorry for the late review.

So, this isn#t that easy. There are two kinds of triggers for actions in logind: edge triggered and level triggered. edge triggered are things like explicit "suspend" or "power" key pressed that happen once and have no duration. level triggered are stuff like the lid switch, i.e. that has a duration: usually a laptop stays with the lid closed for a longer time. Now if we shall suspend the system then for the edge triggered stuff we just issue the suspend once, but for the level triggered stuff we issue the stuff again and again until the condition is over. i.e. when we the lid switch is on, then we keep suspending as long as the lid switch stays on, should we wake up in between. Doing suspends here in a tight loop is fine, given that we talk about system suspend here: our loop will be paused as long as the system is suspended after all. Doing the "lock" stuff for level triggered stuff is more problematic however: locking just means sending out a dbus message after all: if we'd keep doing that while the level triggered stuff is in effect we'd possibly busy loop sending out dbus message, and that's something we cannot do.

The the idle detection logic is a level-triggered condition: once idle the system remains idle for a while: whenever we check again it should come to the same result until eventually idle is over. If we'd keep sending out dbus message asking the sessions to act on it, we'd possibly do this forever, overloading CPU and bus...

The logic here needs to become a bit more complex if we want to open up the screen lock stuff to the idle logic (or any other level triggered triggers, such as the lid switch): we'd have to memorize the previous state, and only issue the action whenever state changes from 0 → 1.  i.e. it would probaby be pretties if the is_edge param to manager_handle_action() would be replaced by an enum that communicates EDGE, LEVEL or FLANK. We'd pass EDGE for one-time events such as explicit key presses. We'd pass FLANK whenever a level-triggered event happens where the state was previously 0 and is 1 now. And we'd pass LEVEL whenever we call this again and the state as previously 1 and is still 1. This would then allow manager_handle_action() to keep issuing the suspend actions on all three cases, but the idle action would only be done on EDGE and FLANK but not on LEVEL, so that we only issue on dbus message whenever the system goes idle and that's.

long story short: in order to not potentially send out unbounded numbers of screen lock dbus signal messages on the bus the patch needs to become a bit smarter and remember state.

**Comment by venkata-pyla** [comment]
thank you @poettering for your detailed explanation, i undestood now the background of the 'is_edge' param in manager_handle_action(), and may be as you suggested we should have some state monitoring and accordingly the 'Lock' signal should send.
I will futher investigate and check if i can contribute here.

Thank you

**Comment by keszybz** [comment]
It's very strange. Github does allow #16391 to be linked as issue for this PR. (And symetrically, it does not allow linking this PR in the issue.)

**Comment by venkata-pyla** [comment]
Thank you for the support for making me understand the complexity of fixing the session Lock.
Currently we are achieving our requirement using with session termination (TMOUT) instead of session lock.
However we are still in need of this feature "session lock" for locking the user text-based login sessions.

Currently I am closing this PR, in future if I got some solution I will raise another PR.
Thank you.
